### ISSUE TITLE
Cifar 10 with Standard classifiers which are available in torchvision.models

### ISSUE THREAD

**Issue by abhiagwl4262** [BayesWatch/nas-without-training#3]
@jack-willturner, So I was trying this methodology with standard classifiers. I am not getting anything useful and I am getting scores as the multiple of batch size. 

**Models chosen** -  
**"alexnet", "resnet18", "resnet34", "resnet50", "vgg11", "vgg13", "densenet121", "squeezenet1_0", "squeezenet1_1"**

 **scores of 3 runs with Batch Size 512** 
[-515.8197304827107, -518.2625130150999, -516.5254794393328, -516.4521991392903, -513.0445545389762, -513.0351409136783, -531.7337148363374, -526.7972858922, -523.973769767945]

[-515.9147741416524, -517.8983410664996, -515.5328530338215, -515.9739907753669, -513.055884116389, -513.0808931741735, -533.8211941483576, -518.7453811789933, -518.6898948738232]

[-515.8902849753072, -518.636564681304, -516.60644909449, -515.057577734371, -513.0407785346342, -513.0632672108508, -527.9074191760042, -518.9160167714085, -521.2741464876093]

**Scores of 3 runs with Batch Size 128 -** 

[-128.26608670187613, -128.9030387211241, -128.50543259689053, -128.45200275464504, -128.0648502228396
5, -128.06361606092634, -131.5393815597839, -130.2799460305505, -129.34274766217487]

[-128.27500716523485, -128.8883337356769, -128.3406425853629, -128.5692902076405, -128.06446535511876,
 -128.06785661838, -132.06972535717804, -128.77940184757026, -128.60743606369488]

[-128.28348337334037, -129.04774963648924, -128.53128977454088, -128.25740881809102, -128.066172268912
1, -128.06880782087086, -130.62699720044384, -128.84707498551833, -128.83386057804955]

**Scores of 3 runs with Batch Size 16 -** 
[-16.00536177981646, -16.023735408663864, -16.00541814906714, -16.01275948633266, -16.001070293337335, -16.000959487618033, -16.088079533081366, -16.00829194903181, -16.008265848684317]

[-16.005564368012045, -16.037812752585545, -16.01270083749682, -16.007351551415354, -16.000882548413536, -16.000910242701053, -16.246709654891255, -16.06660858188009, -16.014516625366127]

[-16.005648553602278, -16.020132359932354, -16.011393278637634, -16.006370556147218, -16.001099750108267, -16.000646158922933, -16.20028038506129, -16.014573365687735, -16.004068435407078]

Any Idea about this ?

**Comment by jack-willturner** [comment]
Could you attach some code so I can take a look?

Thanks!

**Comment by abhiagwl4262** [comment]
@jack-willturner 

This is the code I am using - 

import torch
import torch.nn as nn
import torchvision
import torchvision.transforms as transforms
import argparse
import random
import numpy as np
import os 

parser = argparse.ArgumentParser(description='NAS Without Training')
parser.add_argument('--dataroot', default = './datasets', type = str)
parser.add_argument('--runs', default=3, type=int)
parser.add_argument('--batch_size', default=8, type=int)
parser.add_argument('--seed', default=20, type=int)
args = parser.parse_args()

# Reproducibility
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
random.seed(args.seed)
np.random.seed(args.seed)
torch.manual_seed(args.seed)

order_fn = np.nanargmax

def get_batch_jacobian(net, x, target):
    net.zero_grad()

    x.requires_grad_(True)

    y = net(x)

    y.backward(torch.ones_like(y))
    jacob = x.grad.detach()

    return jacob, target.detach()


def eval_score(jacob, labels=None):
    corrs = np.corrcoef(jacob)
    v, _  = np.linalg.eig(corrs)
    k = 1e-5
    return -np.sum(np.log(v + k) + 1./(v + k))

def main():

    transform_train = transforms.Compose([
        transforms.Resize(224),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
    ])

    # Normalize the test set same as training set without augmentation
    transform_test = transforms.Compose([
        transforms.Resize(224),
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
    ])

    trainset = torchvision.datasets.CIFAR10(
        root=args.dataroot, train=True, download=True, transform=transform_train)
    trainloader = torch.utils.data.DataLoader(
        trainset, batch_size=args.batch_size, shuffle=True, num_workers=2)

    testset = torchvision.datasets.CIFAR10(
        root=args.dataroot, train=False, download=True, transform=transform_test)
    testloader = torch.utils.data.DataLoader(
        testset, batch_size=args.batch_size, shuffle=False, num_workers=2)

    archs = ["alexnet", "resnet18", "resnet34", "resnet50", "vgg11", "vgg13", "densenet121", "squeezenet1_0", "squeezenet1_1"]
    # archs = ["resnet18", "resnet34", "resnet50"]

    for run in range(args.runs):
        scores = []
        for arch in archs:
            net = getattr(torchvision.models, arch)()
            if "classifier" in net._modules.keys():
                if isinstance(net.classifier, nn.Sequential):
                    if isinstance(net.classifier[-1], nn.Linear):
                        net.classifier[-1] = torch.nn.Linear(in_features=net.classifier[-1].in_features, out_features=10)
                else:
                    if isinstance(net.classifier, nn.Linear):
                        net.classifier = torch.nn.Linear(in_features=net.classifier.in_features, out_features=10)
            
            elif "fc" in net._modules.keys():
                if isinstance(net.fc, nn.Sequential):
                    if isinstance(net.fc[-1], nn.Linear):
                        net.fc[-1] = torch.nn.Linear(in_features=net.fc[-1].in_features, out_features=10)
                else:
                    if isinstance(net.fc, nn.Linear):
                        net.fc = torch.nn.Linear(in_features=net.fc.in_features, out_features=10)
                

            # net.fc = torch.nn.Linear(in_features=net.fc.in_features, out_features=10)
            # models.append(net())
            data_iterator = iter(trainloader)
            x, target     = next(data_iterator)
            x, target = x.cuda(), target.cuda()
            net = net.cuda()
            net.eval()
            jacobs, labels= get_batch_jacobian(net, x, target)
            jacobs = jacobs.reshape(jacobs.size(0), -1).cpu().numpy()
            try:
                s = eval_score(jacobs, labels)
            except Exception as e:
                print(e)
                s = np.nan
            scores.append(s)
        print(scores)

        best_arch = archs[order_fn(scores)]
        print(best_arch)

if __name__ == "__main__": 
    
    if not os.path.exists(args.dataroot):
        os.mkdir(args.dataroot)

    main()

**Comment by abhiagwl4262** [comment]
`import torch
import torch.nn as nn
import torchvision
import torchvision.transforms as transforms
import argparse
import random
import numpy as np
import os 

parser = argparse.ArgumentParser(description='NAS Without Training')
parser.add_argument('--dataroot', default = './datasets', type = str)
parser.add_argument('--runs', default=3, type=int)
parser.add_argument('--batch_size', default=8, type=int)
parser.add_argument('--seed', default=20, type=int)
args = parser.parse_args()

# Reproducibility
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
random.seed(args.seed)
np.random.seed(args.seed)
torch.manual_seed(args.seed)

order_fn = np.nanargmax

def get_batch_jacobian(net, x, target):
    net.zero_grad()

    x.requires_grad_(True)

    y = net(x)

    y.backward(torch.ones_like(y))
    jacob = x.grad.detach()

    return jacob, target.detach()


def eval_score(jacob, labels=None):
    corrs = np.corrcoef(jacob)
    v, _  = np.linalg.eig(corrs)
    k = 1e-5
    return -np.sum(np.log(v + k) + 1./(v + k))

def main():

    transform_train = transforms.Compose([
        transforms.Resize(224),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
    ])

    # Normalize the test set same as training set without augmentation
    transform_test = transforms.Compose([
        transforms.Resize(224),
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
    ])

    trainset = torchvision.datasets.CIFAR10(
        root=args.dataroot, train=True, download=True, transform=transform_train)
    trainloader = torch.utils.data.DataLoader(
        trainset, batch_size=args.batch_size, shuffle=True, num_workers=2)

    testset = torchvision.datasets.CIFAR10(
        root=args.dataroot, train=False, download=True, transform=transform_test)
    testloader = torch.utils.data.DataLoader(
        testset, batch_size=args.batch_size, shuffle=False, num_workers=2)

    archs = ["alexnet", "resnet18", "resnet34", "resnet50", "vgg11", "vgg13", "densenet121", "squeezenet1_0", "squeezenet1_1"]
    # archs = ["resnet18", "resnet34", "resnet50"]

    for run in range(args.runs):
        scores = []
        for arch in archs:
            net = getattr(torchvision.models, arch)()
            if "classifier" in net._modules.keys():
                if isinstance(net.classifier, nn.Sequential):
                    if isinstance(net.classifier[-1], nn.Linear):
                        net.classifier[-1] = torch.nn.Linear(in_features=net.classifier[-1].in_features, out_features=10)
                else:
                    if isinstance(net.classifier, nn.Linear):
                        net.classifier = torch.nn.Linear(in_features=net.classifier.in_features, out_features=10)
            
            elif "fc" in net._modules.keys():
                if isinstance(net.fc, nn.Sequential):
                    if isinstance(net.fc[-1], nn.Linear):
                        net.fc[-1] = torch.nn.Linear(in_features=net.fc[-1].in_features, out_features=10)
                else:
                    if isinstance(net.fc, nn.Linear):
                        net.fc = torch.nn.Linear(in_features=net.fc.in_features, out_features=10)
                

            # net.fc = torch.nn.Linear(in_features=net.fc.in_features, out_features=10)
            # models.append(net())
            data_iterator = iter(trainloader)
            x, target     = next(data_iterator)
            x, target = x.cuda(), target.cuda()
            net = net.cuda()
            net.eval()
            jacobs, labels= get_batch_jacobian(net, x, target)
            jacobs = jacobs.reshape(jacobs.size(0), -1).cpu().numpy()
            try:
                s = eval_score(jacobs, labels)
            except Exception as e:
                print(e)
                s = np.nan
            scores.append(s)
        print(scores)

        best_arch = archs[order_fn(scores)]
        print(best_arch)

if __name__ == "__main__": 
    
    if not os.path.exists(args.dataroot):
        os.mkdir(args.dataroot)

    main()`

**Comment by abhiagwl4262** [comment]
@jack-willturner 

Hey You can see the code at https://github.com/abhiagwl4262/NAS_WOT

Thanks.

**Comment by jack-willturner** [comment]
It looks like you’re resizing your images to 224x224 pixels on [this line](https://github.com/abhiagwl4262/NAS_WOT/blob/d21df7942454aa0392b60c329faa62816b170556/main.py#L48) and [this line](https://github.com/abhiagwl4262/NAS_WOT/blob/d21df7942454aa0392b60c329faa62816b170556/main.py#L56) but using CIFAR-10, so you may find that resizing to 32x32 produces more meaningful scores.

Let me know if this helps!

**Comment by abhiagwl4262** [comment]
hey @jack-willturner  Thanks for quick reply. I also doubted the same as the reason for getting supposedly less meaningful scores. I experimented that as well and got the same type of scores.  

You can check the same, I have added the code on my repo. you can run **[main_cifar_32.py](https://github.com/abhiagwl4262/NAS_WOT)** 

**scores with Batch Size 8** 
[-8.012779350631492, -8.010750812358541]
[-8.018866456439536, -8.01511912025244]
[-8.01511063070543, -8.015410845723974]

**scores with Batch Size 128** 
[-131.84361871569075, -131.96096687745427]
[-132.12832095128908, -131.99492007380002]
[-131.97419189882407, -131.84318071457977]

**Comment by abhiagwl4262** [comment]
Hey @jack-willturner, It can turn out to be very good and useful paper for the community. Though I would love to discuss about your idea more, why you thought, something like this can work ?

**Comment by jack-willturner** [comment]
I took at look at this and I think it's expected behaviour.
The eigenvalues are all close to 1 so you should get a number close to your batch size as a result. One thing that is incorrect is the use of `.eval()` on [line 88](https://github.com/abhiagwl4262/NAS_WOT/blob/master/main_cifar_32.py#L88) - this was a mistake in our code and should be removed. Once we update the repo I'll close this issue!

As for why it works - we explained the intuition in Sec 3 of [the paper](https://arxiv.org/pdf/2006.04647.pdf#section.3) but the short version is that how differently a network acts for each datapoint at initialisation gives us clues about how trainable it is (i.e. if it collapses several inputs into one region it may be hard to learn a good classification function).

**Comment by abhiagwl4262** [comment]
Hey @jack-willturner, thanks for looking into this. As per your suggestion, I put models in train mode instead of eval mode. I am still getting the scores with eigen values close to 1. 

I have updated [my repo](https://github.com/abhiagwl4262/NAS_WOT) accordingly. You can check.

**Comment by jack-willturner** [comment]
Your code looks right, and the scores you are getting look right too. We expect the scores to be close to the batch size since this a product of the batch size and a bunch of things that are close to one.

Did the same thing happen for the models you listed above ( "alexnet", "resnet18", "resnet34", "resnet50", "vgg11", "vgg13", "densenet121", "squeezenet1_0", "squeezenet1_1")?

**Comment by abhiagwl4262** [comment]
@jack-willturner  Ohh yeah. In both the cases, it behaved the same way. the main.py will run for  ( "alexnet", "resnet18", "resnet34", "resnet50", "vgg11", "vgg13", "densenet121", "squeezenet1_0", "squeezenet1_1") .

**Comment by jack-willturner** [comment]
OK. As long as it is consistent I am fairly certain this is expected behaviour!
